core/src/com/google/inject/internal/BytecodeGen.java

```diff
@@ -16,412 +16,147 @@
-      @Override
-      public Visibility and(Visibility that) {
-        return this;
-      }
-    };
-
-    public static Visibility forMember(Member member) {
-      if ((member.getModifiers() & (Modifier.PROTECTED | Modifier.PUBLIC)) == 0) {
-        return SAME_PACKAGE;
-      }
-
-      Class[] parameterTypes;
-      if (member instanceof Constructor) {
-        parameterTypes = ((Constructor) member).getParameterTypes();
-      } else {
-        Method method = (Method) member;
-        if (forType(method.getReturnType()) == SAME_PACKAGE) {
-          return SAME_PACKAGE;
-        }
-        parameterTypes = method.getParameterTypes();
-      }
-
-      for (Class<?> type : parameterTypes) {
-        if (forType(type) == SAME_PACKAGE) {
-          return SAME_PACKAGE;
-        }
-      }
-
-      return PUBLIC;
-    }
-
-    public static Visibility forType(Class<?> type) {
-      return (type.getModifiers() & (Modifier.PROTECTED | Modifier.PUBLIC)) != 0
-          ? PUBLIC
-          : SAME_PACKAGE;
+  private static Function<String, BiFunction> fastClass(Executable member) {
+    Class<?> hostClass = member.getDeclaringClass();
+    if (hostClass.getSimpleName().contains(ENHANCER_BY_GUICE_MARKER)) {
+      hostClass = hostClass.getSuperclass();
     }
-
-    public abstract Visibility and(Visibility that);
+    return FAST_CLASSES.get(hostClass);
   }
 
   /**
-   * Loader for Guice-generated classes. For referenced classes, this delegates to either either the
-   * user's classloader (which is the parent of this classloader) or Guice's class loader.
+   * Cache of recent {@link EnhancerBuilder}s.
+   *
+   * <p>Uses weak values so builders can be collected after they're done enhancing.
    */
-  private static class BridgeClassLoader extends ClassLoader {
-
-    BridgeClassLoader() {
-      // use system loader as parent
-    }
-
-    BridgeClassLoader(ClassLoader usersClassLoader) {
-      super(usersClassLoader);
-    }
-
-    @Override
-    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
-
-      if (name.startsWith("sun.reflect") || name.startsWith("jdk.internal.reflect")) {
-        // these reflection classes must be loaded from bootstrap class loader
-        return SystemBridgeHolder.SYSTEM_BRIDGE.classicLoadClass(name, resolve);
-      }
-
-      if (name.startsWith(GUICE_INTERNAL_PACKAGE) || name.startsWith(CGLIB_PACKAGE)) {
-        if (null == GUICE_CLASS_LOADER) {
-          // use special system bridge to load classes from bootstrap class loader
-          return SystemBridgeHolder.SYSTEM_BRIDGE.classicLoadClass(name, resolve);
-        }
-        try {
-          Class<?> clazz = GUICE_CLASS_LOADER.loadClass(name);
-          if (resolve) {
-            resolveClass(clazz);
-          }
-          return clazz;
-        } catch (Throwable e) {
-          // fall-back to classic delegation
+  private static final LoadingCache<Class<?>, EnhancerBuilder> ENHANCER_BUILDERS =
```

Comment:
It could, but would mean the `EnhancerBuilder` was kept around for as long as the host class was loaded. Since this holds a lot of reflective content needed to generate the enhanced class, but not to actually use the enhanced class, then we really want to drop it as soon as we're done building the enhanced class to keep memory use down.

You might wonder why use a weak cache here at all, but testing with various modular applications shows that even this short-lived cache avoids recreating the `EnhancerBuilder` many times during bootstrapping. So this is a compromise between no cache and holding onto the reflected `Method` content for the life of the application.

PS. the various enhancers for a given host class are indeed kept in a `ClassValue`, like the generated fast-classes.

Reply:
Thanks for the explanation!